README.md
# Flutter Callkit Incoming

A Flutter plugin to show incoming call in your Flutter app(Custom for Android/Callkit for iOS).

[![pub package](https://img.shields.io/pub/v/billtech_incoming_call.svg)](https://pub.dev/packages/billtech_incoming_call)
[![Build Status](https://github.com/billtech775/billtech_incoming_call/actions/workflows/main.yml/badge.svg)](https://github.com/billtech775/billtech_incoming_call/actions/workflows/main.yml)

## :star: Features

- Show an incoming call
- Start an outgoing call
- Custom UI Android/Callkit for iOS
- Example using Pushkit/VoIP for iOS

  <br>

## iOS: ONLY WORKING ON REAL DEVICE, not on simulator(Callkit framework not working on simulator)

<br>

## 🚀&nbsp; Installation

1. Install Packages

- Run this command:
  ```console
  flutter pub add billtech_incoming_call
  ```
- Add pubspec.yaml:
  ```console
      dependencies:
        billtech_incoming_call: any
  ```

2. Configure Project

- Android
  - AndroidManifest.xml
  ```
   <manifest...>
       ...
       <!--
           Using for load image from internet
       -->
       <uses-permission android:name="android.permission.INTERNET"/>
   </manifest>
  ```
- iOS
  - Info.plist
  ```
  <key>UIBackgroundModes</key>
  <array>
      <string>processing</string>
      <string>remote-notification</string>
      <string>voip</string>
  </array>
  ```

3. Usage

- Import
  ```console
  import 'package:billtech_incoming_call/billtech_incoming_call.dart';
  ```
- Received an incoming call
  ```dart
    this._currentUuid = _uuid.v4();
    var params = <String, dynamic>{
      'id': _currentUuid,
      'nameCaller': 'Hien Nguyen',
      'appName': 'Callkit',
      'avatar': 'https://i.pravatar.cc/100',
      'handle': '0123456789',
      'type': 0,
      'textAccept': 'Accept',
      'textDecline': 'Decline',
      'textMissedCall': 'Missed call',
      'textCallback': 'Call back',
      'duration': 30000,
      'extra': <String, dynamic>{'userId': '1a2b3c4d'},
      'headers': <String, dynamic>{'apiKey': 'Abc@123!', 'platform': 'flutter'},
      'android': <String, dynamic>{
        'isCustomNotification': true,
        'isShowLogo': false,
        'isShowCallback': false,
        'isShowMissedCallNotification': true,
        'ringtonePath': 'system_ringtone_default',
        'backgroundColor': '#0955fa',
        'backgroundUrl': 'https://i.pravatar.cc/500',
        'actionColor': '#4CAF50'
      },
      'ios': <String, dynamic>{
        'iconName': 'CallKitLogo',
        'handleType': 'generic',
        'supportsVideo': true,
        'maximumCallGroups': 2,
        'maximumCallsPerCallGroup': 1,
        'audioSessionMode': 'default',
        'audioSessionActive': true,
        'audioSessionPreferredSampleRate': 44100.0,
        'audioSessionPreferredIOBufferDuration': 0.005,
        'supportsDTMF': true,
        'supportsHolding': true,
        'supportsGrouping': false,
        'supportsUngrouping': false,
        'ringtonePath': 'system_ringtone_default'
      }
    };
    await BilltechIncomingCall.showCallkitIncoming(params);
  ```
- Show miss call notification

  ```dart
    this._currentUuid = _uuid.v4();
    var params = <String, dynamic>{
      'id': this._currentUuid,
      'nameCaller': 'Hien Nguyen',
      'handle': '0123456789',
      'type': 1,
      'textMissedCall': 'Missed call',
      'textCallback': 'Call back',
      'extra': <String, dynamic>{'userId': '1a2b3c4d'},
    };
    await BilltechIncomingCall.showMissCallNotification(params);
  ```

- Started an outgoing call

  ```dart
    this._currentUuid = _uuid.v4();
    var params = <String, dynamic>{
      'id': this._currentUuid,
      'nameCaller': 'Hien Nguyen',
      'handle': '0123456789',
      'type': 1,
      'extra': <String, dynamic>{'userId': '1a2b3c4d'},
      'ios': <String, dynamic>{'handleType': 'generic'}
    };
    await BilltechIncomingCall.startCall(params);
  ```

- Ended an incoming/outgoing call

  ```dart
    var params = <String, dynamic>{'id': this._currentUuid};
    await BilltechIncomingCall.endCall(params);
  ```

- Ended all calls

  ```dart
    await BilltechIncomingCall.endAllCalls();
  ```

- Get active calls. iOS: return active calls from Callkit, Android: only return last call

  ```dart
    await BilltechIncomingCall.activeCalls();
  ```

  Output

  ```json
  [{"id": "8BAA2B26-47AD-42C1-9197-1D75F662DF78", ...}]
  ```

- Get device push token VoIP. iOS: return deviceToken, Android: Empty

  ```dart
    await BilltechIncomingCall.getDevicePushTokenVoIP();
  ```

  Output

  ```json
  <device token>

  //Example
  d6a77ca80c5f09f87f353cdd328ec8d7d34e92eb108d046c91906f27f54949cd

  ```

  Make sure using `SwiftFlutterCallkitIncomingPlugin.sharedInstance?.setDevicePushTokenVoIP(deviceToken)` inside AppDelegate.swift (<a href="https://github.com/billtech775/billtech_incoming_call/blob/master/example/ios/Runner/AppDelegate.swift">Example</a>)

  ```swift
  func pushRegistry(_ registry: PKPushRegistry, didUpdate credentials: PKPushCredentials, for type: PKPushType) {
      print(credentials.token)
      let deviceToken = credentials.token.map { String(format: "%02x", $0) }.joined()
      //Save deviceToken to your server
      SwiftFlutterCallkitIncomingPlugin.sharedInstance?.setDevicePushTokenVoIP(deviceToken)
  }

  func pushRegistry(_ registry: PKPushRegistry, didInvalidatePushTokenFor type: PKPushType) {
      print("didInvalidatePushTokenFor")
      SwiftFlutterCallkitIncomingPlugin.sharedInstance?.setDevicePushTokenVoIP("")
  }
  ```

- Listen events
  ```dart
    BilltechIncomingCall.onEvent.listen((event) {
      switch (event!.name) {
        case CallEvent.ACTION_CALL_INCOMING:
          // TODO: received an incoming call
          break;
        case CallEvent.ACTION_CALL_START:
          // TODO: started an outgoing call
          // TODO: show screen calling in Flutter
          break;
        case CallEvent.ACTION_CALL_ACCEPT:
          // TODO: accepted an incoming call
          // TODO: show screen calling in Flutter
          break;
        case CallEvent.ACTION_CALL_DECLINE:
          // TODO: declined an incoming call
          break;
        case CallEvent.ACTION_CALL_ENDED:
          // TODO: ended an incoming/outgoing call
          break;
        case CallEvent.ACTION_CALL_TIMEOUT:
          // TODO: missed an incoming call
          break;
        case CallEvent.ACTION_CALL_CALLBACK:
          // TODO: only Android - click action `Call back` from missed call notification
          break;
        case CallEvent.ACTION_CALL_TOGGLE_HOLD:
          // TODO: only iOS
          break;
        case CallEvent.ACTION_CALL_TOGGLE_MUTE:
          // TODO: only iOS
          break;
        case CallEvent.ACTION_CALL_TOGGLE_DMTF:
          // TODO: only iOS
          break;
        case CallEvent.ACTION_CALL_TOGGLE_GROUP:
          // TODO: only iOS
          break;
        case CallEvent.ACTION_CALL_TOGGLE_AUDIO_SESSION:
          // TODO: only iOS
          break;
        case CallEvent.ACTION_DID_UPDATE_DEVICE_PUSH_TOKEN_VOIP:
          // TODO: only iOS
          break;
      }
    });
  ```
- Call from Native (iOS PushKit)

  ```swift
    //Swift
    var info = [String: Any?]()
    info["id"] = "44d915e1-5ff4-4bed-bf13-c423048ec97a"
    info["nameCaller"] = "Hien Nguyen"
    info["handle"] = "0123456789"
    info["type"] = 1
    //... set more data
    SwiftFlutterCallkitIncomingPlugin.sharedInstance?.showCallkitIncoming(flutter_callkit_incoming.Data(args: info), fromPushKit: true)
  ```

  <br>

  ```swift
    //OR
    let data = flutter_callkit_incoming.Data(id: "44d915e1-5ff4-4bed-bf13-c423048ec97a", nameCaller: "Hien Nguyen", handle: "0123456789", type: 0)
    data.nameCaller = "Johnny"
    data.extra = ["user": "abc@123", "platform": "ios"]
    //... set more data
    SwiftFlutterCallkitIncomingPlugin.sharedInstance?.showCallkitIncoming(data, fromPushKit: true)
  ```

  <br>

  ```objc
    //Objective-C
    #if __has_include(<flutter_callkit_incoming/flutter_callkit_incoming-Swift.h>)
    #import <flutter_callkit_incoming/flutter_callkit_incoming-Swift.h>
    #else
    #import "flutter_callkit_incoming-Swift.h"
    #endif

    Data * data = [[Data alloc]initWithId:@"44d915e1-5ff4-4bed-bf13-c423048ec97a" nameCaller:@"Hien Nguyen" handle:@"0123456789" type:1];
    [data setNameCaller:@"Johnny"];
    [data setExtra:@{ @"userId" : @"HelloXXXX", @"key2" : @"value2"}];
    //... set more data
    [SwiftFlutterCallkitIncomingPlugin.sharedInstance showCallkitIncoming:data fromPushKit:YES];
  ```

  <br>

  ```swift
    //send custom event from native
    SwiftFlutterCallkitIncomingPlugin.sharedInstance?.sendEventCustom("customEvent", body: ["customKey": "customValue"])

  ```

4. Properties

   | Prop                 | Description                                                                                                                                                                                            | Default                                                     |
   | -------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ | ----------------------------------------------------------- |
   | **`id`**             | UUID identifier for each call. UUID should be unique for every call and when the call is ended, the same UUID for that call to be used. suggest using <a href='https://pub.dev/packages/uuid'>uuid</a> | Required                                                    |
   | **`nameCaller`**     | Caller's name.                                                                                                                                                                                         | _None_                                                      |
   | **`appName`**        | App's name. using for display inside Callkit(iOS).                                                                                                                                                     | App Name, `Deprecated for iOS > 14, default using App name` |
   | **`avatar`**         | Avatar's URL used for display for Android. `/android/src/main/res/drawable-xxxhdpi/ic_default_avatar.png`                                                                                              | _None_                                                      |
   | **`handle`**         | Phone number/Email/Any.                                                                                                                                                                                | _None_                                                      |
   | **`type`**           | 0 - Audio Call, 1 - Video Call                                                                                                                                                                         | `0`                                                         |
   | **`duration`**       | Incoming call/Outgoing call display time (second). If the time is over, the call will be missed.                                                                                                       | `30000`                                                     |
   | **`textAccept`**     | Text `Accept` used in Android                                                                                                                                                                          | `Accept`                                                    |
   | **`textDecline`**    | Text `Decline` used in Android                                                                                                                                                                         | `Decline`                                                   |
   | **`textMissedCall`** | Text `Missed Call` used in Android (show in miss call notification)                                                                                                                                    | `Missed Call`                                               |
   | **`textCallback`**   | Text `Call back` used in Android (show in miss call notification)                                                                                                                                      | `Call back`                                                 |
   | **`extra`**          | Any data added to the event when received.                                                                                                                                                             | `{}`                                                        |
   | **`headers`**        | Any data for custom header avatar/background image.                                                                                                                                                    | `{}`                                                        |
   | **`android`**        | Android data needed to customize UI.                                                                                                                                                                   | Below                                                       |
   | **`ios`**            | iOS data needed.                                                                                                                                                                                       | Below                                                       |

    <br>

- Android

  | Prop                               | Description                                                                                          | Default                                                           |
  | ---------------------------------- | ---------------------------------------------------------------------------------------------------- | ----------------------------------------------------------------- |
  | **`isCustomNotification`**         | Using custom notifications.                                                                          | `false`                                                           |
  | **`isShowLogo`**                   | Show logo app inside full screen. `/android/src/main/res/drawable-xxxhdpi/ic_logo.png`               | `false`                                                           |
  | **`isShowMissedCallNotification`** | Show missed call notification when timeout                                                           | `true`                                                            |
  | **`isShowCallback`**               | Show callback action from miss call notification.                                                    | `true`                                                            |
  | **`ringtonePath`**                 | File name ringtone. put file into `/android/app/src/main/res/raw/ringtone_default.pm3`               | `system_ringtone_default` <br>using ringtone default of the phone |
  | **`backgroundColor`**              | Incoming call screen background color.                                                               | `#0955fa`                                                         |
  | **`backgroundUrl`**                | Using image background for Incoming call screen. example: http://... https://... or "assets/abc.png" | _None_                                                            |
  | **`actionColor`**                  | Color used in button/text on notification.                                                           | `#4CAF50`                                                         |

    <br>

- iOS

  | Prop                                        | Description                                                                                       | Default                                                                                                                      |
  | ------------------------------------------- | ------------------------------------------------------------------------------------------------- | ---------------------------------------------------------------------------------------------------------------------------- |
  | **`iconName`**                              | App's Icon. using for display inside Callkit(iOS)                                                 | `CallKitLogo` <br> using from `Images.xcassets/CallKitLogo`                                                                  |
  | **`handleType`**                            | Type handle call `generic`, `number`, `email`                                                     | `generic`                                                                                                                    |
  | **`supportsVideo`**                         |                                                                                                   | `true`                                                                                                                       |
  | **`maximumCallGroups`**                     |                                                                                                   | `2`                                                                                                                          |
  | **`maximumCallsPerCallGroup`**              |                                                                                                   | `1`                                                                                                                          |
  | **`audioSessionMode`**                      |                                                                                                   | _None_, `gameChat`, `measurement`, `moviePlayback`, `spokenAudio`, `videoChat`, `videoRecording`, `voiceChat`, `voicePrompt` |
  | **`audioSessionActive`**                    |                                                                                                   | `true`                                                                                                                       |
  | **`audioSessionPreferredSampleRate`**       |                                                                                                   | `44100.0`                                                                                                                    |
  | **`audioSessionPreferredIOBufferDuration`** |                                                                                                   | `0.005`                                                                                                                      |
  | **`supportsDTMF`**                          |                                                                                                   | `true`                                                                                                                       |
  | **`supportsHolding`**                       |                                                                                                   | `true`                                                                                                                       |
  | **`supportsGrouping`**                      |                                                                                                   | `true`                                                                                                                       |
  | **`supportsUngrouping`**                    |                                                                                                   | `true`                                                                                                                       |
  | **`ringtonePath`**                          | Add file to root project xcode `/ios/Runner/Ringtone.caf` and Copy Bundle Resources(Build Phases) | `Ringtone.caf`<br>`system_ringtone_default` <br>using ringtone default of the phone                                          |

5. Source code

   ```
   please checkout repo github
   https://github.com/hiennguyen92/flutter_callkit_incoming
   ```

   - <a href='https://github.com/billtech775/billtech_incoming_call'>https://github.com/billtech775/billtech_incoming_call</a>
   - <a href='https://github.com/billtech775/billtech_incoming_call/blob/master/example/lib/main.dart'>Example</a>

  <br>

6. Pushkit - Received VoIP and Wake app from Terminated State (only for IOS)

- Please check <a href="https://github.com/billtech775/billtech_incoming_call/blob/master/PUSHKIT.md">PUSHKIT.md</a> setup Pushkit for IOS

  <br>

7. Todo

-
- <br>

## :bulb: Demo

1. Demo Illustration:
2. Image
<table>
  <tr>
    <td>iOS(Lockscreen)</td>
    <td>iOS(full screen)</td>
    <td>iOS(Alert)</td>
  </tr>
  <tr>
    <td>
      <img src="https://raw.githubusercontent.com/billtech775/billtech_incoming_call/master/images/image1.png" width="220">
    </td>
    <td>
      <img src="https://raw.githubusercontent.com/billtech775/billtech_incoming_call/master/images/image2.png" width="220">
    </td>
    <td>
      <img src="https://raw.githubusercontent.com/billtech775/billtech_incoming_call/master/images/image3.png" width="220">
    </td>
  </tr>
  <tr>
    <td>Android(Lockscreen) - Audio</td>
    <td>Android(Alert) - Audio</td>
    <td>Android(Lockscreen) - Video</td>
  </tr>
  <tr>
    <td>
      <img src="https://raw.githubusercontent.com/billtech775/billtech_incoming_call/master/images/image4.jpg" width="220">
    </td>
    <td>
      <img src="https://raw.githubusercontent.com/billtech775/billtech_incoming_call/master/images/image5.jpg" width="220">
    </td>
    <td>
      <img src="https://raw.githubusercontent.com/billtech775/billtech_incoming_call/master/images/image6.jpg" width="220">
    </td>
  </tr>
  <tr>
    <td>Android(Alert) - Video</td>
    <td>isCustomNotification: false</td>
    <td></td>
  </tr>
  <tr>
    <td>
      <img src="https://raw.githubusercontent.com/billtech775/billtech_incoming_call/master/images/image7.jpg" width="220">
    </td>
    <td>
      <img src="https://raw.githubusercontent.com/billtech775/billtech_incoming_call/master/images/image8.jpg" width="220">
    </td>
  </tr>
 </table>


CHANGELOG.md
## 1.0.0+2

- News icons.

## 1.0.0+1

- News icons.

## 1.0.0+0

- Initial release.


example/lib/main.dart
import 'package:firebase_core/firebase_core.dart';
import 'package:flutter/material.dart';
import 'package:flutter_callkit_incoming_example/app_router.dart';
import 'package:flutter_callkit_incoming_example/navigation_service.dart';
import 'dart:async';
import 'dart:convert';
import 'package:uuid/uuid.dart';
import 'package:firebase_messaging/firebase_messaging.dart';
import 'package:billtech_incoming_call/billtech_incoming_call.dart';

Future<void> _firebaseMessagingBackgroundHandler(RemoteMessage message) async {
  print("Handling a background message: ${message.messageId}");
  showCallkitIncoming(Uuid().v4());
}

Future<void> showCallkitIncoming(String uuid) async {
  var params = <String, dynamic>{
    'id': uuid,
    'nameCaller': 'Hien Nguyen',
    'appName': 'Callkit',
    'avatar': 'https://i.pravatar.cc/100',
    'handle': '0123456789',
    'type': 0,
    'duration': 30000,
    'textAccept': 'Accept',
    'textDecline': 'Decline',
    'textMissedCall': 'Missed call',
    'textCallback': 'Call back',
    'extra': <String, dynamic>{'userId': '1a2b3c4d'},
    'headers': <String, dynamic>{'apiKey': 'Abc@123!', 'platform': 'flutter'},
    'android': <String, dynamic>{
      'isCustomNotification': true,
      'isShowLogo': false,
      'isShowCallback': false,
      'ringtonePath': 'system_ringtone_default',
      'backgroundColor': '#0955fa',
      'backgroundUrl': 'https://i.pravatar.cc/500',
      'actionColor': '#4CAF50'
    },
    'ios': <String, dynamic>{
      'iconName': 'CallKitLogo',
      'handleType': '',
      'supportsVideo': true,
      'maximumCallGroups': 2,
      'maximumCallsPerCallGroup': 1,
      'audioSessionMode': 'default',
      'audioSessionActive': true,
      'audioSessionPreferredSampleRate': 44100.0,
      'audioSessionPreferredIOBufferDuration': 0.005,
      'supportsDTMF': true,
      'supportsHolding': true,
      'supportsGrouping': false,
      'supportsUngrouping': false,
      'ringtonePath': 'system_ringtone_default'
    }
  };
  await BilltechIncomingCall.showCallkitIncoming(params);
}

void main() {
  WidgetsFlutterBinding.ensureInitialized();
  runApp(MyApp());
}

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> with WidgetsBindingObserver {
  var _uuid;
  var _currentUuid;

  late final FirebaseMessaging _firebaseMessaging;

  @override
  void initState() {
    super.initState();
    _uuid = Uuid();
    initFirebase();
    WidgetsBinding.instance?.addObserver(this);
    //Check call when open app from terminated
    checkAndNavigationCallingPage();
  }

  getCurrentCall() async {
    //check current call from pushkit if possible
    var calls = await BilltechIncomingCall.activeCalls();
    if (calls is List) {
      if (calls.isNotEmpty) {
        print('DATA: $calls');
        this._currentUuid = calls[0]['id'];
        return calls[0];
      } else {
        this._currentUuid = "";
        return null;
      }
    }
  }

  checkAndNavigationCallingPage() async {
    var currentCall = await getCurrentCall();
    if (currentCall != null) {
      NavigationService.instance
          .pushNamedIfNotCurrent(AppRoute.callingPage, args: currentCall);
    }
  }

  @override
  Future<void> didChangeAppLifecycleState(AppLifecycleState state) async {
    print(state);
    if (state == AppLifecycleState.resumed) {
      //Check call when open app from background
      checkAndNavigationCallingPage();
    }
  }

  @override
  void dispose() {
    WidgetsBinding.instance?.removeObserver(this);
    super.dispose();
  }

  initFirebase() async {
    await Firebase.initializeApp();
    _firebaseMessaging = FirebaseMessaging.instance;
    FirebaseMessaging.onBackgroundMessage(_firebaseMessagingBackgroundHandler);
    FirebaseMessaging.onMessage.listen((RemoteMessage message) async {
      print(
          'Message title: ${message.notification?.title}, body: ${message.notification?.body}, data: ${message.data}');
      this._currentUuid = _uuid.v4();
      showCallkitIncoming(this._currentUuid);
    });
    _firebaseMessaging.getToken().then((token) {
      print('Device Token FCM: $token');
    });
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      theme: ThemeData.light(),
      onGenerateRoute: AppRoute.generateRoute,
      initialRoute: AppRoute.homePage,
      navigatorKey: NavigationService.instance.navigationKey,
      navigatorObservers: <NavigatorObserver>[
        NavigationService.instance.routeObserver
      ],
    );
  }

  Future<void> getDevicePushTokenVoIP() async {
    var devicePushTokenVoIP =
        await BilltechIncomingCall.getDevicePushTokenVoIP();
    print(devicePushTokenVoIP);
  }
}


example/lib/home_page.dart
import 'package:flutter/material.dart';
import 'package:billtech_incoming_call/billtech_incoming_call.dart';
import 'package:flutter_callkit_incoming_example/app_router.dart';
import 'package:flutter_callkit_incoming_example/navigation_service.dart';
import 'package:uuid/uuid.dart';
import 'dart:async';
import 'dart:convert';
import 'package:http/http.dart';

class HomePage extends StatefulWidget {
  @override
  State<StatefulWidget> createState() {
    return HomePageState();
  }
}

class HomePageState extends State<HomePage> {
  var _uuid;
  var _currentUuid;
  var textEvents = "";

  @override
  void initState() {
    super.initState();
    _uuid = Uuid();
    _currentUuid = "";
    textEvents = "";
    initCurrentCall();
    listenerEvent(onEvent);
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('Plugin example app'),
        actions: <Widget>[
          IconButton(
            icon: Icon(
              Icons.call,
              color: Colors.white,
            ),
            onPressed: () async {
              this.makeFakeCallInComing();
            },
          ),
          IconButton(
            icon: Icon(
              Icons.call_end,
              color: Colors.white,
            ),
            onPressed: () async {
              this.endCurrentCall();
            },
          ),
          IconButton(
            icon: Icon(
              Icons.call_made,
              color: Colors.white,
            ),
            onPressed: () async {
              this.startOutGoingCall();
            },
          ),
          IconButton(
            icon: Icon(
              Icons.call_merge,
              color: Colors.white,
            ),
            onPressed: () async {
              this.activeCalls();
            },
          ),
          IconButton(
            icon: Icon(
              Icons.clear_all_sharp,
              color: Colors.white,
            ),
            onPressed: () async {
              this.endAllCalls();
            },
          )
        ],
      ),
      body: LayoutBuilder(
        builder: (BuildContext context, BoxConstraints viewportConstraints) {
          if (textEvents.isNotEmpty) {
            return SingleChildScrollView(
              child: ConstrainedBox(
                constraints: BoxConstraints(
                  minHeight: viewportConstraints.maxHeight,
                ),
                child: Text('$textEvents'),
              ),
            );
          } else {
            return Center(
              child: Text('No Event'),
            );
          }
        },
      ),
    );
  }

  initCurrentCall() async {
    //check current call from pushkit if possible
    var calls = await BilltechIncomingCall.activeCalls();
    if (calls is List) {
      if (calls.isNotEmpty) {
        print('DATA: $calls');
        this._currentUuid = calls[0]['id'];
        return calls[0];
      } else {
        this._currentUuid = "";
        return null;
      }
    }
  }

  Future<void> makeFakeCallInComing() async {
    await Future.delayed(const Duration(seconds: 10), () async {
      this._currentUuid = _uuid.v4();
      var params = <String, dynamic>{
        'id': _currentUuid,
        'nameCaller': 'Hien Nguyen',
        'appName': 'Callkit',
        'avatar': 'https://i.pravatar.cc/100',
        'handle': '0123456789',
        'type': 0,
        'duration': 30000,
        'textAccept': 'Accept',
        'textDecline': 'Decline',
        'textMissedCall': 'Missed call',
        'textCallback': 'Call back',
        'extra': <String, dynamic>{'userId': '1a2b3c4d'},
        'headers': <String, dynamic>{
          'apiKey': 'Abc@123!',
          'platform': 'flutter'
        },
        'android': <String, dynamic>{
          'isCustomNotification': true,
          'isShowLogo': false,
          'isShowCallback': true,
          'isShowMissedCallNotification': true,
          'ringtonePath': 'system_ringtone_default',
          'backgroundColor': '#0955fa',
          'backgroundUrl': 'assets/test.png',
          'actionColor': '#4CAF50'
        },
        'ios': <String, dynamic>{
          'iconName': 'CallKitLogo',
          'handleType': '',
          'supportsVideo': true,
          'maximumCallGroups': 2,
          'maximumCallsPerCallGroup': 1,
          'audioSessionMode': 'default',
          'audioSessionActive': true,
          'audioSessionPreferredSampleRate': 44100.0,
          'audioSessionPreferredIOBufferDuration': 0.005,
          'supportsDTMF': true,
          'supportsHolding': true,
          'supportsGrouping': false,
          'supportsUngrouping': false,
          'ringtonePath': 'system_ringtone_default'
        }
      };
      await BilltechIncomingCall.showCallkitIncoming(params);
    });
  }

  Future<void> endCurrentCall() async {
    initCurrentCall();
    var params = <String, dynamic>{'id': this._currentUuid};
    await BilltechIncomingCall.endCall(params);
  }

  Future<void> startOutGoingCall() async {
    this._currentUuid = _uuid.v4();
    var params = <String, dynamic>{
      'id': this._currentUuid,
      'nameCaller': 'Hien Nguyen',
      'handle': '0123456789',
      'type': 1,
      'extra': <String, dynamic>{'userId': '1a2b3c4d'},
      'ios': <String, dynamic>{'handleType': 'number'}
    }; //number/email
    await BilltechIncomingCall.startCall(params);
  }

  Future<void> activeCalls() async {
    var calls = await BilltechIncomingCall.activeCalls();
    print(calls);
  }

  Future<void> endAllCalls() async {
    await BilltechIncomingCall.endAllCalls();
  }

  Future<void> getDevicePushTokenVoIP() async {
    var devicePushTokenVoIP =
        await BilltechIncomingCall.getDevicePushTokenVoIP();
    print(devicePushTokenVoIP);
  }

  Future<void> listenerEvent(Function? callback) async {
    try {
      BilltechIncomingCall.onEvent.listen((event) async {
        print('HOME: $event');
        switch (event!.name) {
          case CallEvent.ACTION_CALL_INCOMING:
            // TODO: received an incoming call
            break;
          case CallEvent.ACTION_CALL_START:
            // TODO: started an outgoing call
            // TODO: show screen calling in Flutter
            break;
          case CallEvent.ACTION_CALL_ACCEPT:
            // TODO: accepted an incoming call
            // TODO: show screen calling in Flutter
            NavigationService.instance
                .pushNamedIfNotCurrent(AppRoute.callingPage, args: event.body);
            break;
          case CallEvent.ACTION_CALL_DECLINE:
            // TODO: declined an incoming call
            await requestHttp("ACTION_CALL_DECLINE_FROM_DART");
            break;
          case CallEvent.ACTION_CALL_ENDED:
            // TODO: ended an incoming/outgoing call
            break;
          case CallEvent.ACTION_CALL_TIMEOUT:
            // TODO: missed an incoming call
            break;
          case CallEvent.ACTION_CALL_CALLBACK:
            // TODO: only Android - click action `Call back` from missed call notification
            break;
          case CallEvent.ACTION_CALL_TOGGLE_HOLD:
            // TODO: only iOS
            break;
          case CallEvent.ACTION_CALL_TOGGLE_MUTE:
            // TODO: only iOS
            break;
          case CallEvent.ACTION_CALL_TOGGLE_DMTF:
            // TODO: only iOS
            break;
          case CallEvent.ACTION_CALL_TOGGLE_GROUP:
            // TODO: only iOS
            break;
          case CallEvent.ACTION_CALL_TOGGLE_AUDIO_SESSION:
            // TODO: only iOS
            break;
          case CallEvent.ACTION_DID_UPDATE_DEVICE_PUSH_TOKEN_VOIP:
            // TODO: only iOS
            break;
        }
        if (callback != null) {
          callback(event.toString());
        }
      });
    } on Exception {}
  }

  //check with https://webhook.site/#!/2748bc41-8599-4093-b8ad-93fd328f1cd2
  Future<void> requestHttp(content) async {
    get(Uri.parse(
        'https://webhook.site/2748bc41-8599-4093-b8ad-93fd328f1cd2?data=$content'));
  }

  onEvent(event) {
    if (!mounted) return;
    setState(() {
      textEvents += "${event.toString()}\n";
    });
  }
}
